### ISSUE TITLE
Confusion matrix calculation 

### ISSUE THREAD

**Issue by saskra** [talboger/fastdist#4]
Thank you for this great package, it speeds up my scripts tremendously!

However, I noticed in the following test that the speed increase is "only" six times (from 2.5 h to 24 min). 

```python
import time

import numpy as np
from fastdist import fastdist
from sklearn import metrics

start = time.time()
y_true = np.random.randint(2, size=10000000000).astype(np.uint8)
y_pred = np.random.randint(2, size=10000000000).astype(np.uint8)
end = time.time()
print(end - start)

start = time.time()
cm1 = fastdist.confusion_matrix(y_true, y_pred)
end = time.time()
print(end - start)

start = time.time()
cm2 = metrics.confusion_matrix(y_true, y_pred)
end = time.time()
print(end - start)
```

In your last example here https://github.com/talboger/fastdist/blob/master/README.md#speed you gave a 97-fold increase for a smaller example to calculate the accuracy. Is this because this particular function is simply much more optimised by you compared to the one in the same sklearn.metrics package? Does this also apply to Balanced Accuracy, Dice/ F1, Jaccard and MCC? With sklearn, the confusion matrix is recalculated each time, so I have switched to calculating it only once and then determining the four values with my own formulas. With your package, would it be faster to calculate the values directly or to take the diversions via the previous calculation of the confusion matrix? (If you don't know this yet, I'll do a test on this myself when I get a chance).

**Comment by talboger** [comment]
Great question! Some functions happen to be much more numba-optimized than others, making the fastdist speed-up much larger. I have not personally compared numba vs. sklearn.metrics for all these functions, but I assume their speed-ups will be closer to the accuracy score speed-up than the confusion matrix speed-up (will check to be sure).

That's a great point about recalculating the confusion matrices. Our balanced accuracy, recall, etc. each recalculate the confusion matrices within the function, so yes it'll be faster to calculate it once and then use your own formulas. I think this could actually result in a giant speed-up, so will implement this in a future version (i.e., allowing the functions to take in a confusion matrix so it doesn't have to get recalculated each time).

**Comment by saskra** [comment]
> I think this could actually result in a giant speed-up, so will implement this in a future version (i.e., allowing the functions to take in a confusion matrix so it doesn't have to get recalculated each time).

I would really appreciate this improvement!

Another feature request: Would it be feasible for the function to also work with lists?

**Comment by talboger** [comment]
numba has very strict types which makes it generally inconvenient to also work with both lists and arrays. I'll see if there's another workaround that can help, but in general I'd bet the best thing is to just convert them to numpy arrays.

**Comment by saskra** [comment]
Thank you, then I will do the conversion myself.

By the way, I noticed something else about your implementation: If I have two classes but they are not called 0 and 1, they are not counted, see example below. The background is that I wanted to follow the convention that the "healthy" belong to the first class and the "sick" to the second class, because the order of the majority and minority class has an impact on the calculation of F1 and Jaccard/ IoU.

```python
y_true = np.random.randint(2, 4, size=1000).astype(np.uint8)
y_pred = np.random.randint(2, 4, size=1000).astype(np.uint8)
```

**Comment by talboger** [comment]
Yes that's correct. For speed reasons it's much more convenient to assume that the classes start at 0. In your case it's probably best to just map the healthy/sick arrays to 0s and 1s as you would like with a dictionary (or whatever else is faster).

**Comment by saskra** [comment]
OK, then I'll add that to my private conversion function too, so it remains compatible with sklearn. (And I'll convert the output to uint64.) Perhaps meaningful warnings about unsupported input formats would be a nice service?

**Comment by talboger** [comment]
Yes will clarify that a bit more in the future - generally just a big numba thing that once you compile a function then it only accepts parameters of the same type of the initial compile (i.e., if you compile an accuracy score with a vector of 0s and 1s, then it will throw an error if you give it vectors of strings). Because of this, the defaults are what most people will do (in this case, the 0s and 1s), so yes I should clarify.

Anyway, I've published v1.1.1 where balanced accuracy, recall, etc. allow you to pass in an already-calculated confusion matrix, resulting in significant speed ups. One important note which is explained in the parameters of recall and precision is that certain recall/precision types require certain adjusted confusion matrices (average = 'macro' or 'none' in recall require a confusion matrix with normalize = 'true', and these averages require normalize = 'pred' in precision), so make sure the confusion matrix is created with those necessary parameters. Let me know if you have any issues with this or any additional suggestions!
